File: sdk/reproit-backend-rb/lib/reproit_backend_rb/instrument.rb
# Outbound-exchange capture and hermetic replay for reproit-backend-rb.
#
# `Instrument.install` prepends a module to `Net::HTTP#request`, so every
# dependency call made while a request trace is ambient is recorded on that
# trace as an `effect` event carrying an `exchange`: the request the app sent
# and the response the dependency returned. Because virtually every Ruby HTTP
# client (Net::HTTP itself, HTTParty, Faraday's default adapter, Octokit)
# bottoms out in `Net::HTTP#request`, that one hook is automatic for the
# common case. Clients that open raw sockets or use libcurl bindings are
# invisible to it; those record through `Instrument.db`-style explicit calls
# or not at all, which is stated rather than papered over.
#
# With `REPROIT_REPLAY` naming a capture payload the SAME hook serves the
# recorded exchanges: no socket is opened, so replay needs no live
# dependency. Every capture path fails closed the other way: an
# instrumentation defect must never break the host application's request.

require "net/http"

require_relative "exchange"
require_relative "replay"

module ReproitBackendRb
  module Instrument
    TRACE_KEY = :reproit_backend_rb_trace

    @lock = Mutex.new
    @installed = false
    @session = nil
    @session_loaded = false
    @stats = { captured_exchanges: 0, truncated_bodies: 0, failed_captures: 0 }

    module_function

    # The ambient trace for the current thread, or nil. Framework adapters
    # scope it around the handler; `Instrument.db` and the Net::HTTP hook
    # read it. Thread-local (fiber-local in Ruby), which is the scope a Rack
    # handler runs in.
    def current_trace
      trace = Thread.current[TRACE_KEY]
      trace if trace && !trace.finished?
    end

    def with_trace(trace)
      previous = Thread.current[TRACE_KEY]
      Thread.current[TRACE_KEY] = trace
      yield
    ensure
      Thread.current[TRACE_KEY] = previous
    end

    # Load the replay session once. Also pins the process envelope (TZ), so
    # calling `install` from the entry point pins it before any time-zone
    # sensitive code runs.
    def session
      @lock.synchronize do
        unless @session_loaded
          @session_loaded = true
          path = ENV["REPROIT_REPLAY"].to_s
          unless path.strip.empty?
            @session = Replay::Session.load(path)
            Replay.pin_envelope(@session.envelope)
          end
        end
        @session
      end
    end

    def replaying?
      !session.nil?
    end

    # The capture's seeded stream, or nil outside replay mode. Documented
    # honestly: this pins REPLAY determinism, not the app's original draws.
    def replay_rng
      handle = session
      handle.nil? ? nil : Replay.rng_for(handle.envelope)
    end

    def stats
      @lock.synchronize { @stats.dup }
    end

    def count(key)
      @lock.synchronize { @stats[key] += 1 }
    end

    # Install the Net::HTTP hook once, process-wide. Idempotent.
    def install
      @lock.synchronize do
        return false if @installed
        Net::HTTP.prepend(NetHttpHook)
        @installed = true
      end
      session
      true
    end

    def installed?
      @lock.synchronize { @installed }
    end

    # Record one exchange on the ambient trace. Shared by the HTTP hook and
    # the database helper; never raises into the host.
    def record(kind, resource, key, exchange)
      trace = current_trace
      return if trace.nil?
      trace.effect(kind, resource: resource, key: key, exchange: exchange)
      count(:captured_exchanges)
    rescue StandardError
      count(:failed_captures)
      nil
    end

    # Generic database boundary: run `block` and record the statement with
    # its result, or serve the recorded result in replay mode without
    # touching a driver. The block is never called while replaying, which is
    # what makes a replay run valid with the database stopped.
    #
    # `block` returns a hash shaped { "rows" => [...], "command" => ..,
    # "rowCount" => .. }; anything else records as a bare row count.
    def db(text, values = nil)
      handle = session
      probe = { "text" => text.to_s }
      probe["values"] = values if values.is_a?(Array) && !values.empty?
      unless handle.nil?
        recorded = handle.match("db", probe)
        raise DivergenceError, "reproit: database call diverged from the capture" if recorded.nil?
        outcome = recorded["response"] || {}
        if outcome["error"]
          raise RecordedError, outcome["error"]["message"].to_s
        end
        return {
          "command" => outcome["command"],
          "rowCount" => outcome["rowCount"] || 0,
          "rows" => outcome["rows"].is_a?(Array) ? outcome["rows"] : [],
        }
      end
      begin
        result = yield
      rescue StandardError => error
        record(
          Exchange.statement_effect_kind(text), "db", text.to_s[0, 256],
          Exchange.db(text, values, Exchange.db_error(error))
        )
        raise
      end
      record(
        Exchange.statement_effect_kind(text), "db", text.to_s[0, 256],
        Exchange.db(text, values, Exchange.db_outcome(result))
      )
      result
    end

    # Raised when a replayed call has no matching recorded exchange.
    class DivergenceError < StandardError; end

    # Raised in replay mode when the capture recorded a driver error.
    class RecordedError < StandardError; end

    # Synthesize a Net::HTTPResponse from a served exchange. No socket is
    # involved, so the caller's code path is identical to a live response.
    def synthesize_response(served)
      code = served["status"].to_i
      klass = Net::HTTPResponse::CODE_TO_OBJ[code.to_s] || Net::HTTPResponse
      message = code == 599 ? "Reproit Diverged" : "OK"
      response = klass.new("1.1", code.to_s, message)
      served["headers"].each { |name, value| response[name.to_s] = value.to_s }
      response.instance_variable_set(:@body, served["body"])
      response.instance_variable_set(:@read, true)
      chunks = served["chunks"]
      if chunks.is_a?(Array)
        # Recorded stream shape (SSE / chunked): a block-consuming reader
        # observes the recorded chunk boundaries, chunk for chunk.
        response.define_singleton_method(:read_body) do |dest = nil, &chunk_block|
          if chunk_block
            chunks.each { |chunk| chunk_block.call(chunk) }
            @body
          elsif dest
            chunks.each { |chunk| dest << chunk }
            @body
          else
            @body
          end
        end
      end
      response
    end

    # Prepended to Net::HTTP so one hook covers every client built on it.
    module NetHttpHook
      def request(request_object, body = nil, &block)
        session = Instrument.session
        unless session.nil?
          return Instrument.__replay_request(self, request_object, body, &block)
        end
        trace = Instrument.current_trace
        return super unless trace
        if block
          # Streaming consumer (`request { |res| res.read_body { ... } }`,
          # the SSE shape): recording TEES the chunks the app itself
          # consumes, so observed chunk boundaries are preserved and the
          # exchange lands at EOF of the body, never by draining a stream
          # the app still needs. A body the app abandons is drained by
          # Net::HTTP itself after the block, so it records as one chunk.
          return super(request_object, body) do |response|
            begin
              Instrument.__attach_tee(self, request_object, body, response, trace)
            rescue StandardError
              Instrument.count(:failed_captures)
            end
            block.call(response)
          end
        end
        response = super
        begin
          Instrument.__record_request(self, request_object, body, response)
        rescue StandardError
          Instrument.count(:failed_captures)
        end
        response
      end

      private

      # `Net::HTTP.get_response` and friends open the socket in `start`,
      # BEFORE `request` runs, so hooking `request` alone would still resolve
      # DNS and connect. In replay mode the connection is skipped entirely:
      # the session serves every response, so no socket is ever needed and a
      # replay run is valid with the network denied.
      def connect
        return if Instrument.replaying?
        super
      end
    end

    def absolute_url(http, request_object)
      scheme = http.respond_to?(:use_ssl?) && http.use_ssl? ? "https" : "http"
      port = http.port
      default = scheme == "https" ? 443 : 80
      authority = port == default ? http.address.to_s : "#{http.address}:#{port}"
      "#{scheme}://#{authority}#{request_object.path}"
    rescue StandardError
      request_object.path.to_s
    end

    def request_headers(request_object)
      headers = {}
      request_object.each_header { |name, value| headers[name.to_s] = value }
      headers
    end

    def response_headers(response)
      headers = {}
      response.each_header { |name, value| headers[name.to_s] = value }
      headers
    end

    # Internal: record one live Net::HTTP exchange.
    def __record_request(http, request_object, body, response)
      # A streaming response consumed by a caller block leaves `body` nil;
      # the exchange is recorded without content rather than half-guessed.
      response_body = begin
        response.body
      rescue StandardError
        nil
      end
      content_type = response["content-type"].to_s
      # This path drains in one read, so the observed stream shape is
      # coarse: SSE records its boundaries as one drained chunk. Fine
      # boundaries live on the tee path (a block-consuming caller).
      stream = nil
      if content_type.include?("text/event-stream") && response_body.is_a?(String) &&
         !response_body.empty?
        collector = Exchange::BodyCollector.new
        collector.push(response_body)
        stream = collector.stream(true)
      end
      __record_exchange(http, request_object, body, response, response_body, stream)
    end

    # Internal: the shared tail of both capture paths.
    def __record_exchange(http, request_object, body, response, response_body, stream)
      request_body = body || request_object.body
      exchange = Exchange.http(
        {
          method: request_object.method.to_s.upcase,
          url: absolute_url(http, request_object),
          headers: request_headers(request_object),
          body: request_body,
          content_type: request_object["content-type"].to_s,
        },
        {
          status: response.code.to_i,
          headers: response_headers(response),
          body: response_body,
          content_type: response["content-type"].to_s,
          stream: stream,
        }
      )
      count(:truncated_bodies) if exchange["response"]["truncated"] == true
      record("call", http.address.to_s, "#{request_object.method} #{request_object.path}", exchange)
    end

    # Internal: tee a block-consumed response body. The prepended singleton
    # module records at EOF with the chunk boundaries the app observed; a
    # body that is never read past records nothing beyond what Net::HTTP
    # itself drains after the caller's block.
    def __attach_tee(http, request_object, body, response, trace)
      collector = Exchange::BodyCollector.new
      content_type = response["content-type"].to_s
      finish = lambda do
        stream = collector.stream(content_type.include?("text/event-stream"))
        with_trace(trace) do
          __record_exchange(http, request_object, body, response, collector.result, stream)
        end
      end
      tee = Module.new do
        define_method(:read_body) do |dest = nil, &chunk_block|
          if instance_variable_get(:@reproit_teed)
            next chunk_block ? super(dest, &chunk_block) : super(dest)
          end
          instance_variable_set(:@reproit_teed, true)
          result =
            if chunk_block
              super(dest) do |chunk|
                begin
                  collector.push(chunk)
                rescue StandardError
                  Instrument.count(:failed_captures)
                end
                chunk_block.call(chunk)
              end
            else
              drained = super(dest)
              collector.push(drained) if dest.nil? && drained.is_a?(String)
              drained
            end
          # EOF: the app has seen the whole body; record exactly once.
          begin
            finish.call
          rescue StandardError
            Instrument.count(:failed_captures)
          end
          result
        end
      end
      response.singleton_class.prepend(tee)
      nil
    end

    # Internal: serve one recorded exchange, opening no socket. Mirrors
    # Net::HTTP#request's block form: the synthesized response is yielded.
    def __replay_request(http, request_object, body)
      handle = session
      request_body = body || request_object.body
      probe = {
        "method" => request_object.method.to_s.upcase,
        "url" => absolute_url(http, request_object),
      }
      unless request_body.nil? || request_body.to_s.empty?
        probe["body"] = Replay.try_json(request_body.to_s, request_object["content-type"].to_s)
      end
      response = synthesize_response(Replay.serve_http(handle, probe))
      yield response if block_given?
      response
    end
  end
end

File: sdk/reproit-backend-php/instrument.php
<?php

/*!
 * Outbound-exchange capture and hermetic replay for reproit-backend-php.
 *
 * PHP cannot monkeypatch a client the way Node can, so the boundary is
 * explicit and OPT-IN, exactly like the Rust SDK: route outbound HTTP
 * through `Instrument::http()` (or wrap a PSR-18 client with
 * `RecordingClient`) and database statements through `Instrument::db()`, and
 * every dependency exchange (request AND response) is recorded onto the
 * ambient request trace, bounded and redacted at source.
 *
 * What IS automatic once the boundary is used: the ambient trace comes from
 * the request-scoped singleton the framework adapters set, so handlers do
 * not thread a trace through their own code, and with `REPROIT_REPLAY` set
 * the SAME entry points serve the recorded exchanges with no socket and no
 * driver. Anything bypassing the boundary (a raw curl_exec, an ORM with its
 * own connection) is invisible to capture and unavailable at replay; that is
 * stated rather than papered over.
 *
 * Every capture path fails closed the other way: an instrumentation defect
 * must never break the host application's request.
 */

declare(strict_types=1);

namespace ReproitBackend;

require_once __DIR__ . '/exchange.php';
require_once __DIR__ . '/replay.php';
require_once __DIR__ . '/trace.php';

/** Raised when a replayed call has no matching recorded exchange. */
final class DivergenceError extends \RuntimeException
{
}

/** Raised in replay mode when the capture recorded a driver error. */
final class RecordedError extends \RuntimeException
{
}

/** The served side of one HTTP exchange, whether live or replayed. */
final class HttpResponse
{
    public function __construct(
        public readonly int $status,
        /** @var array<string, string> */
        public readonly array $headers,
        public readonly string $body,
    ) {
    }

    public function json(): mixed
    {
        return json_decode($this->body, true);
    }
}

final class Instrument
{
    private static ?BackendTrace $trace = null;
    private static ?ReplaySession $session = null;
    private static bool $sessionLoaded = false;
    private static array $stats = [
        'capturedExchanges' => 0,
        'truncatedBodies' => 0,
        'failedCaptures' => 0,
    ];

    /**
     * The request-scoped ambient trace. PHP serves one request per process
     * in the classic model, so a static singleton IS the request scope; the
     * framework adapters set and clear it around the handler.
     */
    public static function setTrace(?BackendTrace $trace): void
    {
        self::$trace = $trace;
    }

    public static function currentTrace(): ?BackendTrace
    {
        if (self::$trace === null || self::$trace->finished()) {
            return null;
        }
        return self::$trace;
    }

    /**
     * Load the replay session once and pin the process envelope. Calling
     * this from the entry point pins the timezone before any date-sensitive
     * code runs.
     */
    public static function session(): ?ReplaySession
    {
        if (!self::$sessionLoaded) {
            self::$sessionLoaded = true;
            $path = (string) (getenv('REPROIT_REPLAY') ?: '');
            if (trim($path) !== '') {
                self::$session = ReplaySession::load($path);
                pin_envelope(self::$session->envelope());
            }
        }
        return self::$session;
    }

    public static function replaying(): bool
    {
        return self::session() !== null;
    }

    /** The capture's seeded stream, or null outside replay mode. */
    public static function replayRng(): ?ReplayRng
    {
        $session = self::session();
        return $session === null ? null : rng_for($session->envelope());
    }

    /**
     * The capture's wall-clock instant in epoch milliseconds, or null outside
     * replay mode. PHP cannot pin the process clock without an extension (see
     * pin_envelope in replay.php for the measured evidence), so an app that
     * needs an anchored clock in replay reads this and uses it as its time
     * source instead of calling time() directly.
     */
    public static function replayObservedAtMs(): ?int
    {
        $session = self::session();
        return $session === null ? null : observed_at_ms($session->envelope());
    }

    /**
     * The SDK clock seam: a PinnedClock offset to the capture instant in
     * replay mode, the system clock otherwise. PHP cannot intercept time()
     * process wide without an extension (pin_envelope documents the measured
     * evidence), so an app that wants replay-anchored time reads THIS clock
     * instead of the ambient one; no mode branch needed in app code.
     */
    public static function clock(): Clock
    {
        $observed = self::replayObservedAtMs();
        return $observed === null ? new SystemClock() : new PinnedClock($observed);
    }

    /** @return array<string, int> */
    public static function stats(): array
    {
        return self::$stats;
    }

    public static function count(string $key): void
    {
        self::$stats[$key] = (self::$stats[$key] ?? 0) + 1;
    }

    /** Record one exchange on the ambient trace; never raises into the host. */
    public static function record(string $kind, string $resource, string $key, array $exchange): void
    {
        try {
            $trace = self::currentTrace();
            if ($trace === null) {
                return;
            }
            $trace->effect($kind, [
                'resource' => $resource,
                'key' => $key,
                'exchange' => $exchange,
            ]);
            self::count('capturedExchanges');
        } catch (\Throwable) {
            self::count('failedCaptures');
        }
    }

    /**
     * Outbound HTTP boundary. Sends the request with stdlib streams and
     * records the exchange, or serves the recorded one in replay mode
     * without opening a socket.
     *
     * @param array<string, string> $headers
     */
    public static function http(
        string $method,
        string $url,
        ?string $body = null,
        array $headers = [],
        float $timeoutSeconds = 5.0
    ): HttpResponse {
        $method = strtoupper($method);
        $contentType = '';
        foreach ($headers as $name => $value) {
            if (strtolower((string) $name) === 'content-type') {
                $contentType = (string) $value;
            }
        }
        $session = self::session();
        if ($session !== null) {
            $probe = ['method' => $method, 'url' => $url];
            if ($body !== null && $body !== '') {
                $probe['body'] = try_json($body, $contentType);
            }
            $served = serve_http($session, $probe);
            return new HttpResponse($served['status'], $served['headers'], $served['body']);
        }
        $headerLines = '';
        foreach ($headers as $name => $value) {
            $headerLines .= $name . ': ' . $value . "\r\n";
        }
        $context = stream_context_create(['http' => [
            'method' => $method,
            'header' => $headerLines,
            'content' => $body ?? '',
            'timeout' => $timeoutSeconds,
            'ignore_errors' => true,
        ]]);
        $raw = @file_get_contents($url, false, $context);
        $responseHeaders = [];
        $status = 0;
        foreach ($http_response_header ?? [] as $line) {
            if (preg_match('#^HTTP/\S+\s+(\d{3})#', $line, $matches) === 1) {
                $status = (int) $matches[1];
                continue;
            }
            $split = explode(':', $line, 2);
            if (\count($split) === 2) {
                $responseHeaders[strtolower(trim($split[0]))] = trim($split[1]);
            }
        }
        $responseBody = $raw === false ? '' : $raw;
        try {
            self::record('call', parse_url($url, PHP_URL_HOST) ?: 'http', $method . ' ' . $url, http_exchange(
                [
                    'method' => $method,
                    'url' => $url,
                    'headers' => $headers,
                    'body' => $body,
                    'contentType' => $contentType,
                ],
                [
                    'status' => $status,
                    'headers' => $responseHeaders,
                    'body' => $responseBody,
                    'contentType' => $responseHeaders['content-type'] ?? '',
                ]
            ));
        } catch (\Throwable) {
            self::count('failedCaptures');
        }
        return new HttpResponse($status, $responseHeaders, $responseBody);
    }

    /**
     * Generic database boundary: run `$run` and record the statement with
     * its result, or serve the recorded result in replay mode without
     * touching a driver. `$run` is never called while replaying, which is
     * what makes a replay run valid with the database stopped.
     *
     * `$run` returns ['rows' => [...], 'command' => .., 'rowCount' => ..].
     */
    public static function db(string $text, ?array $values, callable $run): array
    {
        $session = self::session();
        if ($session !== null) {
            $probe = ['text' => $text];
            if ($values !== null && $values !== []) {
                $probe['values'] = array_values($values);
            }
            $recorded = $session->match('db', $probe);
            if ($recorded === null) {
                throw new DivergenceError('reproit: database call diverged from the capture');
            }
            $outcome = $recorded['response'] ?? [];
            if (isset($outcome['error'])) {
                throw new RecordedError((string) ($outcome['error']['message'] ?? 'recorded error'));
            }
            return [
                'command' => $outcome['command'] ?? null,
                'rowCount' => $outcome['rowCount'] ?? 0,
                'rows' => \is_array($outcome['rows'] ?? null) ? $outcome['rows'] : [],
            ];
        }
        try {
            $result = $run();
        } catch (\Throwable $error) {
            self::record(
                statement_effect_kind($text),
                'db',
                substr($text, 0, 256),
                db_exchange($text, $values, db_error($error))
            );
            throw $error;
        }
        self::record(
            statement_effect_kind($text),
            'db',
            substr($text, 0, 256),
            db_exchange($text, $values, db_outcome($result))
        );
        return \is_array($result) ? $result : ['rows' => [], 'rowCount' => 0];
    }

    /** Test seam: forget the loaded session so a new REPROIT_REPLAY applies. */
    public static function resetForTests(): void
    {
        self::$session = null;
        self::$sessionLoaded = false;
        self::$trace = null;
    }
}
